### DISCUSSION TITLE
My-Linux-Config/docs/nvim

### DISCUSSION THREAD

**Discussion by giscus** [Martins3/My-Linux-config#65] [General]
# My-Linux-Config/docs/nvim

:clap: Modern neovim configuration based on coc.nvim

https://martins3.github.io/My-Linux-Config/docs/nvim.html

**Comment by zhehaoli1999** [comment]
Such a nice article! Respect and marked.

**Reply by Martins3** [reply]
Thank you, and suggestion is welcomed :yum:

**Comment by hao-lee** [comment]
最开始用ctags，后来用gtags，现在又想换更好的，正好搜到这篇文章，太强了！！！除了这三个字我没有其他形容词。

**Comment by astralwaveorg** [comment]
Linux使用了，很舒服，是否可以出一份 macOS（ apple silicon ） 适用的配置呢

**Reply by IfanTsai** [reply]
我就同时也在 macbook pro 上使用，暂时没有遇到过跟 Linux 表现不一致的现象，理论上只要 neovim 支持的系统就都可以 （在 16 寸的 intel 版本和 m1 pro 上都使用过）

**Comment by txctxc** [comment]
Nice article! Respect and marked.
记录并和大家分享一下我在Ubuntu22.04上面按照教程配置中教程没有详细叙述的步骤：
Ubuntu22.04安装LSP(**Language Server Protocol** ):

1. [marksman](https://github.com/artempyanykh/marksman/blob/main/docs/install.md)
2. [efm](https://github.com/mattn/efm-langserver#installation)
3. [ccls](https://github.com/MaskRay/ccls/wiki/Install)

```
sudo apt update
apt list --upgradable
sudo apt upgrade
sudo apt -y install efm-langserver
sudo snap install marksman
sudo apt -y install ccls
sudo apt install npm
sudo apt -y install cargo
sudo apt -y install xclip
sudo apt install lua5.4

mv ~/.config/nvim ~/.config/nvim.bak
cd ~
export https_proxy=http://127.0.0.1:7890 && export http_proxy=http://127.0.0.1:7890 && export all_proxy=socks5://127.0.0.1:7890
git clone --depth=1 https://github.com/martins3/My-Linux-config .dotfiles
ln -s ~/.dotfiles/nvim ~/.config/nvim
nvim
## 在 nvim 中间执行 checkhealth 命令
## 在 nvim 中间执行 :Lazy build rsync.nvim命令

用nvim打开一个.c或者.cpp文件，在neovim中执行命令：
：LspInfo
查看Lsp状态。
下面是怎么使用ccls建立一个项目，参考https://github.com/MaskRay/ccls/wiki/LanguageClient-neovim 

cd ~/.config/nvim
vim setting.json
```

在setting.json中写入以下内容：

```
{
  "initializationOptions": {
    "cache": {
      "directory": "/home/t/NutstoreFiles/Nutstore/cppcodes/ccls"
    }
  }
}
```

vim ~/.config/nvim/init.vim

在init.vim中写入以下内容：

```
let g:LanguageClient_serverCommands = {
    \ 'c': ['ccls', '--log-file=/cppcodes/cc.log'],
    \ 'cpp': ['ccls', '--log-file=/cppcodes/cc.log'],
    \ 'cuda': ['ccls', '--log-file=/cppcodes/cc.log'],
    \ 'objc': ['ccls', '--log-file=/cppcodes/cc.log'],
    \ }

let g:LanguageClient_loadSettings = 1 " Use an absolute configuration path if you want system-wide settings
let g:LanguageClient_settingsPath = '/home/t/.config/nvim/settings.json'
" https://github.com/autozimu/LanguageClient-neovim/issues/379 LSP snippet is not supported
"let g:LanguageClient_hasSnippetSupport = 0
```

配置ccls的项目建立的第一个例子：

```R
cd /home/t/NutstoreFiles/Nutstore/cppcodes; mkdir c; cd c
git init  # let LanguageClient-neovim know this is a project
echo '#include <stddef.h>\nint main() {}' > a.cc
nvim a.cc
# :LanguageClientStart
```

For a larger project, you'll need [.ccls](https://github.com/MaskRay/ccls/wiki/Project-Setup#ccls-file) or [compile_commands.json](https://github.com/MaskRay/ccls/wiki/Project-Setup#compile_commandsjson) in your project root.

**Reply by Martins3** [reply]
从这里开始，这部分内容是不对的

> 查看Lsp状态。
下面是怎么使用ccls建立一个项目，参考https://github.com/MaskRay/ccls/wiki/LanguageClient-neovim 

我们现在用的是 native lsp ，无需额外的使用 LanguageClient-neovim ，这些内容不加也是可以正常运行。

**Comment by ravenxrz** [comment]
写得真不错

**Comment by Hehe72-lang** [comment]
大神 我在用m1 mac air配置的时候一直报一个错:
   33 lua require 'usr'
这一行报
[im-select]: binary tools missed, please follow installation manual in README

我本机是安装了lua的，很奇怪 网上也没找到解决方案

**Reply by Martins3** [reply]
这是因为 im-select 的依赖没有安装，提一个 issue ，我们在 issue 中讨论。

**Comment by Eta-Chu** [comment]
大佬，请问一下我安装了clangd，但是在写的时候，不能识别cpp文件之间的link关系，也不能识别补全外部库的函数，这是哪里出了问题？多谢🙏

**Reply by Martins3** [reply]
有 compile_comands.json 吗？最好还是提个 issue 描述清楚吧

**Comment by dawushi97** [comment]
我想知道有没有在头文件选择声明，快速生成定义的方法，类似Clion中的生成定义，用LSP吗还是需要自定义？

**Reply by Martins3** [reply]
应该是没有的

**Reply by dawushi97** [reply]
好的,谢谢答复。
